References
----------

.. [Angerer16] Angerer *et al.* (2016),
   *destiny â€“ diffusion maps for large-scale single-cell data in R*,
   `Bioinformatics <https://doi.org/10.1093/bioinformatics/btv715>`__.
   
.. [Cusanovich18] Cusanovich, D. A. et al. (2018),
	*A Single-Cell Atlas of In Vivo Mammalian Chromatin Accessibility*,
	`Cell <https://pubmed.ncbi.nlm.nih.gov/30078704/>`__.
   
.. [Danese21] Danese, A. et al. (2021),
	*EpiScanpy: integrated single-cell epigenomic analysis*,
	`Nature Communications <https://www.nature.com/articles/s41467-021-25131-3>`__.

.. [Luo17] Luo, C. et al. (2017),
	*Single-cell methylomes identify neuronal subtypes and regulatory elements in mammalian cortex.*,
	`Science <https://www.science.org/doi/abs/10.1126/science.aan3351>`__.
   
.. [Wolf18] Wolf, F. A., Angerer, P. & Theis, F. J. (2018),
	*SCANPY: large-scale single-cell gene expression data analysis*,
	`Genome Biology <https://genomebiology.biomedcentral.com/articles/10.1186/s13059-017-1382-0>`__.

.. [Buenrostro18] Buenrostro J. D. et al. (2018),
	*Integrated Single-Cell Analysis Maps the Continuous Regulatory Landscape of Human Hematopoietic Differentiation*,
	`Cell <https://pubmed.ncbi.nlm.nih.gov/29706549/>`__.